### DISCUSSION TITLE
Problem running the MNIST example

### DISCUSSION THREAD

**Discussion by yhsure** [e3nn/e3nn#344] [Q&A]
Hello, 

I am working on a regression task much related to the the MNIST example.
However, when I run examples/s2cnn/mnist/train.py I cannot seem to reproduce the S2CNN results. 
While the network is slowly learning, I only go from ~20% test accuracy in epoch 1 to ~40% in epoch 20.
Other learning rates etc did not seem to help. Do you have any idea about what to change? 

Also thanks for the great work!

**Comment by mariogeiger** [comment]
I still need to investigate this

**Comment by dmklee** [comment]
I was also getting a performance of about 40% accuracy after 20 epochs.  I think it may have something to do with the order of alpha and beta.  The original s2cnn repo uses beta then alpha in [s2_near_identity_grid](https://github.com/jonkhler/s2cnn/blob/b75efee458686e7d7ecb4c337402c668ede0dece/s2cnn/s2_grid.py#L5)) whereas it looks like e3nn uses alpha then beta in its version of [s2_near_identity_grid](https://github.com/e3nn/e3nn/blob/40c4e143a664bc1b63921f7f7c4f26f73821af69/examples/s2cnn/mnist/train.py#L11)?  When I corrected for this by adding `x = x.transpose(-1,-2)` to the beginning of `S2ConvNet_original.forward`, it achieves an accuracy of ~85% after 20 epochs.

This is still less than the 95% accuracy reported in the paper.  I noticed that the degree of spherical harmonics is explicitly defined in the e3nn MNIST example, whereas in s2cnn it is set to the bandwidth of the incoming signal (see [here](https://github.com/jonkhler/s2cnn/blob/b75efee458686e7d7ecb4c337402c668ede0dece/s2cnn/soft/s2_fft.py#L82)).  I tried increasing `lmax1` but this did not improve performance.  

I want to use spherical convolution for a project, so I will keep looking into this.  Any guidance would be much appreciated

**Reply by mariogeiger** [reply]
Thanks for looking into the details of the code, I'm sorry that I didn't have time to do it myself. 
Would you like to make a PR that fixes the issue you found?

**Reply by mariogeiger** [reply]
I just added that fix https://github.com/e3nn/e3nn/commit/0cce011e7483c7323d529cbb51108e336fa6f8cb thanks again!
The missing 10% of accuracy might also come from the fact that the SO3 transform is less precise in this code.

**Reply by dmklee** [reply]
Thanks for making the fix!  I have gotten up to 92% by removing the normalizing of the input images (94% with `lmax1=15`).  In the current implementation there is `train_data /= 57` and `test_data /= 57`, which is not present in the original implementation.  Strangely, normalizing the input signals seems to hurt performance, which may be due to the information loss when converting the signal to spherical harmonics of a limited degree.

**Reply by mariogeiger** [reply]
Very interesting point, I like the hypothesis, sounds very plausible! Let me remove it in the code
